Map Block: Add zoom and height settings controls

pento opened the pull request:
(Note: This PR is currently based off the branch for #14673, since that includes some changes that this one needs.)

This PR introduces sidebar settings for map zoom and height control.

The zoom control is a reflection of existing ability to set the zoom by scrolling over the map.

The height control is new, and also uses `<ResizableBox>`, which adds a clickable handle to the bottom of the map, allowing it to be dragged and resized.

Finally, this PR tweaks the zoom behaviour when points are added: when there are 2 or more points on the map, zoom customisations are already ignored, so the zoom UI is now disabled to match.

Fixes #14557.

#### Changes proposed in this Pull Request:
* Map Block: Add zoom and height controls.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Enhancement to the Map Block.

#### Testing instructions:

* Open the block editor.
* Add a new map block.
* Try setting the height/zoom with the sidebar control, and check that the block updates.
* Try setting the height with the resize handle, and check that everything updates.
* Try setting the zoom by scrolling on the map, and check that the sidebar control updates.

#### Proposed changelog entry for your changes:
* Map Block: Add the ability to set the size of the map.
* Map Block: Add a zoom control to the block sidebar.

pento committed:
Add map height and zoom settings.

simison commented on extensions/blocks/map/edit.js:
```diff
@@ -185,6 +212,51 @@ class MapEdit extends Component {
+									setAttributes( {
+										mapHeight: height,
+									} );
+									setTimeout( this.mapRef.current.sizeMap, 0 );
```
You could do this in `componentDidUpdate`, [example](https://github.com/Automattic/jetpack/blob/bf52e73459946042c4e587a2acbe9f1d643b6a4e/extensions/blocks/tiled-gallery/layout/mosaic/index.js#L29) more cleanly than using timeout.

pento replied:
I copied this from the existing usage, which comes from when the map block was originally built (fc7596995ad368bbc6f88aa35afe0ef3ede18433). Reading through the original PRs (Automattic/wp-calypso#27836, Automattic/wp-calypso#28498), I couldn't see any discussion of why it was done this way.

@jeffersonrabb: I know it was a long time ago, but I don't suppose you recall if there was a specific reason for the `setTimeout()` call here, rather than using `componentDidUpdate()`?

jeffersonrabb replied:
> @jeffersonrabb: I know it was a long time ago, but I don't suppose you recall if there was a specific reason for the `setTimeout()` call here, rather than using `componentDidUpdate()`?

No idea 😬. I added the line in the initial commit so there's no useful commit message. Given that there's no `componentDidUpdate` in `MapEdit` it's probably safe to assume the approach just didn't occur to me!

pento replied:
I'm going to leave it as is for now, since it'd be a bit weird to only fix some, but not this others, and this PR has been going on for long enough as is. 🙂

simison commented on extensions/blocks/map/component.js:
```diff
@@ -177,18 +177,30 @@ export class Map extends Component {
-		const blockHeight = Math.min( blockWidth * ( 3 / 4 ), maxHeight );
-		mapEl.style.height = blockHeight + 'px';
+		if ( mapHeight ) {
+			mapEl.style.height = mapHeight + 'px';
```
This seems odd for React; I'd expect to see `style={ { height: ... } }` in JSX instead.

simison replied:
Oh I see it's what was already done before in the block: https://github.com/Automattic/jetpack/pull/14706/files#diff-fc2a014531c8c68c34dcc4c8292c5f21L188

pento replied:
Yah, I left that as is, I assumed that the direct manipulation of `mapEl` is deliberate.

simison commented on extensions/blocks/map/edit.js:
```diff
@@ -185,6 +212,51 @@ class MapEdit extends Component {
+							<input
+								type="number"
+								id={ `block-jetpack-map-height-input-${ instanceId }` }
+								onChange={ event => {
```
These functions are little lengthy within JSX, I'd consider moving the out and referencing them. 🙂

pento replied:
👍🏻 91e5b20bd

mtias commented on extensions/blocks/map/edit.js:
```diff
@@ -271,40 +343,65 @@ class MapEdit extends Component {
-						markerColor={ markerColor }
-						onSetZoom={ value => {
-							setAttributes( { zoom: value } );
+					<ResizableBox
```
Nice to see `ResizableBox` in use :)

Copons commented on extensions/blocks/map/edit.js:
```diff
@@ -271,40 +343,65 @@ class MapEdit extends Component {
+							height: mapHeight || 'auto',
+							width: '100%',
+						} }
+						showHandle={ isSelected }
```
Just pointing out that this is a newer prop, not available in WP yet, but the component works perfectly fine without it (just, without the handle, obviously).

pento replied:
Good thing to note. For reference, it was added in WordPress/gutenberg#18097. Prior to that, adding the `is-selected` class to the `ResizableBox` component was the way to make it appear, d9aab4dfa adds this as a fallback.

Copons commented on extensions/blocks/map/edit.js:
```diff
@@ -185,6 +212,51 @@ class MapEdit extends Component {
+									} );
+									setTimeout( this.mapRef.current.sizeMap, 0 );
+								} }
+								value={ mapHeight || '' }
```
I've noticed that the default map height is `420`.
The first time you click on either of this input's arrows when the value is empty, the value becomes `400` regardless, and the map size shrinks a little bit.

![Feb-17-2020 12-36-12](https://user-images.githubusercontent.com/2070010/74654406-206f8800-5182-11ea-9046-dde220d8f2cd.gif)

I find it a bit weird, especially when clicking on the increase arrow.

What about something like this?

```jsx
const DEFAULT_HEIGHT = 420;
const MIN_HEIGHT = 400;

<input
	value={ mapHeight || DEFAULT_HEIGHT }
	onChange={ event => {
		let height = parseInt( event.target.value, 10 );
		if ( isNaN( height ) || DEFAULT_HEIGHT === height ) {
			height = null;
		} else if ( height < MIN_HEIGHT ) {
			height = MIN_HEIGHT;
		}
		setAttributes( { mapHeight: height } );
		setTimeout( this.mapRef.current.sizeMap, 0 );
	} }
/>
```

pento replied:
The default map height changes based on the window width: you can see this by resizing the editor window.

1dae92a52 uses the current map height, however it's calculated, as the value to insert when changing the height from default, using the incr/decr buttons. There are a couple of slightly weird things about this:

- Clicking the incr/decr buttons for the first time won't change the height of the map. I think this works to show the post author what they've done, so I'm not too concerned about it. 
- The first value it shows may not be divisible by 10, which means the next incr/decr click will just round it up/down. Slightly weird, but I don't think it's a deal breaker.

pento replied:
Also worth noting: The `ResizableBox` handle snaps to a 10px grid: the first move of it will snap it to the grid, additional resizing will move in 10px increments/decrements.

jeherve replied:
I came here to mention something similar; I got confused when I first clicked on the field and it went from empty to a default as soon as I tried typing something. I wanted to enter `600`, but it ended up entering `43500` because it switched to `435` when I typed `6`, and then carried on with the 0s I added.

When I saw that, I tried to hit backspace to remove everything and type 600 again, but I could not do that; I only allowed me to delete to 400. 

As it is, I don't know how to set the height to 600, short of using the little arrows.

pento replied:
Ugh, I forgot that React triggers the `change` event when it receives an `input` event.

Fixed in 28cdc95eb.

Copons commented on extensions/blocks/map/edit.js:
```diff
@@ -185,6 +212,51 @@ class MapEdit extends Component {
+									setTimeout( this.mapRef.current.sizeMap, 0 );
+								} }
+								value={ mapHeight || '' }
+								min="400"
```
We are using this value quite a lot across the block.
What about using a constant instead? (also used in my [previous comment](https://github.com/Automattic/jetpack/pull/14706/files?utf8=%E2%9C%93&diff=split&w=1#r380161712) 🙂)
```js
const MIN_HEIGHT = 400;
```

pento replied:
👍🏻 f0924df3f

Copons commented on extensions/blocks/map/edit.js:
```diff
@@ -185,6 +212,51 @@ class MapEdit extends Component {
+											'The default zoom level cannot be changed when there are two or more markers on the map.',
+											'jetpack'
+									  )
+									: ''
```
We might simplify this as `points.length > 1 && __( 'The default...' )`

pento replied:
👍🏻 07b5cc0aa

pento replied:
👍🏻 07b5cc0aa

Copons commented:
This is working really well for me! ✨

pento committed:
Define the minimum map height as a constant.

pento committed:
If there was no height defined when using the incr/decr buttons to change the height, set it to the default height.

pento committed:
Simplify the zoom control help text logic.

pento committed:
Add a fallback for ResizableBox to show the resize handle in older WP versions.

pento committed:
Refactor the event handlers out of the JSX.

pento committed:
Add docblocks for the new event handlers.

pento committed:
Automatically hide the excess map when using the ResizeBox handler to make the map smaller.

pento committed:
Use the MapBox placeholder grey to fill the additional space when making the map larger, until the map can re-render.

pento commented:
I've addressed all the feedback on the PR to date, except [this question](https://github.com/Automattic/jetpack/pull/14706#discussion_r380022707).

There are a couple of noteworthy behaviour changes today:
- The `ResizableBox` handler now locks to a 10px grid, to match the behaviour of the height input control.
- When using the `ResizableBox` handler to make the map smaller, it will automatically hide the excess map (it would previously overflow, which looked kind of janky). When making it larger, the excess space will be filled with the same grey that MapBox uses as a placeholder before map tiles are loaded.

jeherve commented on extensions/blocks/map/edit.js:
```diff
@@ -185,6 +270,37 @@ class MapEdit extends Component {
+							/>
+						</BaseControl>
+						<RangeControl
+							label={ __( 'Zoom level', 'jetpack' ) }
```
Maybe we could rely on `withInputField` and hide the number field when zoom is disabled?

![image](https://user-images.githubusercontent.com/426388/74763883-e507b300-5280-11ea-9770-89819b9f158f.png)

pento replied:
I'm not sure what `withInputField` is, `RangeControl` [doesn't have an option](https://github.com/WordPress/gutenberg/blob/5c926f8f67e76c26eed5415ff0360ed06d5c1764/packages/components/src/range-control/index.js#L93-L103) for hiding the number field, that I can see.

jeherve replied:
My bad, I should have checked the source. I was going off the documentation here:
https://developer.wordpress.org/block-editor/components/range-control/#withinputfield

But it seems it was only added a few days ago, so that's not something we'll be able to use right away:
https://github.com/WordPress/gutenberg/pull/19916

pento replied:
Huh, cool. I'm inclined to not use it yet, since relatively few sites will have access to it, so it'll just cause JS console warnings about an unknown prop being passed on to the DOM element.

jeherve commented on extensions/blocks/map/edit.js:
```diff
@@ -185,6 +270,37 @@ class MapEdit extends Component {
+							label={ __( 'Height in pixels' ) }
+							id={ `block-jetpack-map-height-input-${ instanceId }` }
+						>
+							<input
```
I wonder if we could style this a bit better (this is in WP 5.3):

![image](https://user-images.githubusercontent.com/426388/74763644-77f41d80-5280-11ea-8c1f-2b7197756d97.png)

It looks better in 5.2 though:

![image](https://user-images.githubusercontent.com/426388/74765023-1aad9b80-5283-11ea-9619-a187ec25c68e.png)

pento replied:
👍🏻 8ea23b5a6

pento committed:
Add a missing translation textdomain.

pento committed:
Keep the height input box on its own line.

pento committed:
Check the correctness of the height input on blur, so it doesn't interfere with typing.

pento commented on extensions/blocks/map/edit.js:
```diff
@@ -185,6 +270,37 @@ class MapEdit extends Component {
 					/>
 					<PanelBody title={ __( 'Map Settings', 'jetpack' ) }>
+						<BaseControl
+							label={ __( 'Height in pixels' ) }
```
👍🏻 379157e70

jeherve reviewed (approved):
This tests well for me now. 👍
